design/feature-flags.md

```diff
@@ -0,0 +1,136 @@
+
+In order for new projects to pick up this new behavior automatically, we will
+modify `cdk init` to inject the set of feature flags into the generated
+`cdk.json` file. This means that the new project will have the latest behavior,
+but projects that were created prior to the introduction of this feature will
+have the same legacy behavior based on the set of capabilities that were
+available at the time of the project's creation. This list will be cleaned up
+every time we release a major version of course.
+
+Using fine-grained flags will allow users of old projects to pick up specific
+new behaviors by manually adding the specific keys to their `cdk.json`
+file, without risking breakage in other unexpected areas.
+
+## Alternative Considered
+ 
+We considered an alternative of "bundling" new capabilities under a single flag
+that specifies the CDK version which created the project, but this means that
+users won't have the ability to pick and choose which capabilities they
+want to enable in case they need them but don't want to take the risk of
+unexpected changes.
+ 
+The downside of the fine-grained approach is that it could result in a "blowing
+up" new `cdk.json` files in case there will be many new breaking capabilities
+between major releases. But this is hypothetical and even if this list ends up
+with 20 features before we release the next major version, I still think the
+benefits outweigh the risks of the alternative approach.
+
+## Design
+
+Context keys for feature flags will be listed in `cx-api/lib/features.ts` and
+will take the form: `<module>:<feature>`. 
+
+For example:
+
+- `@aws-cdk/core:enableStackNameDuplicates`
+- `@aws-cdk/aws-cloudformation:doNotCapitalizeCustomResourcePropertyNames`.
+
+Using the module name will allow easy tracing of the code that consumes this
+flag.
+
+The configuration for which feature flags should be enabled for new projects
+will be under `cx-api/lib/future.ts` and will be encoded as a simple context
+hash that will be injected by `cdk init` to all `cdk.json` files generated for
+new projects.
+
+We will mandate that when a feature or bug fix is introduced under a feature
+flag, the CHANGELOG will include:
+
+- The suffix `(under feature flag)` in the title.
+- A `BREAKING CHANGES` paragraph will be added which describes the *new*
+  behavior but disclaims that it will only apply to new projects created through
+  `cdk init`. It will also indicate the context key this flag uses for users who
+  wish to enable it manually in their project.
+
+Since feature flags can have implications on framework behavior, we need to
+ask users to include the list of enabled features in bug reports. At a minimum,
+we can request that they paste a copy of their `cdk.json` and `cdk.context.json`,
+but a better experience would be to include this information in the output of
+`cdk doctor` and request users to include this output in bug reports.
+
+## Developer Experience
```

Comment:
This section is telling the store of the construct author's experience. We should also design the app builder's experience.

Feature flags as a concept needs to be documented somewhere centrally [here](https://docs.aws.amazon.com/cdk/latest/guide/core_concepts.html) which includes explanations around how to find the list of feature flags their app is using and how to toggle them.

Besides this, there needs to be a single place to enumerate the list of feature flags we have, when to use them and what changes are needed to use them.

Reply:
Yes, we will add this to the docs. About listing feature flags, initially we will just point people to the source code. If we see that this becomes too confusing, we can consider adding cli commands for listing, enabling etc. I don’t want to over engineer this especially given we want this to be kept for very rare cases.